### PR TITLE
add print dialog to work with user's custom template

### PR DESCRIPTION
**Pull request by myarjunar**
### What does it fix?
* Ticket: #4523 #4240 
* Funded by: DFAT
* Description: Add a print dialog for custom map report purpose.
- [x] have to fix the legend when trying to generate report from print dialog
- [x] pop up all pdf products (just like before)
- [x] fix the text
- [x] fix item properties when trying to open in map composer
 
selection_009

### Checklist:
- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [x] Request someone to review or test your PR

### CODE REVIEW

File: safe/definitions/utilities.py [hunk 1]
```diff
@@ -485,6 +487,67 @@ def get_field_groups(layer_purpose, layer_subcategory=None):
...
+    # we need to update the orientation of the custom template
+    with open(custom_template_component['template']) as (
+            template_file):
+        template_content = template_file.read()
+    document = QtXml.QDomDocument()
+    document.setContent(template_content)
+    root_element = document.namedItem('Composer')
+    orientation = None
+    if isinstance(root_element, QDomNode):
+        title_atribute_node = root_element.attributes().namedItem('title')
```
Comment (Gustry) [inline safe/definitions/utilities.py]: `attribute`

File: safe/gui/tools/print_report_dialog.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+FORM_CLASS = get_ui_class('print_report_dialog.ui')
+LOGGER = logging.getLogger('InaSAFE')
+
+
+class PrintReportDialog(QtGui.QDialog, FORM_CLASS):
+    """Print report dialog for the InaSAFE plugin."""
+
+    def __init__(self, impact_function, iface, dock=None, parent=None):
+        """Constructor for the dialog.
+        :param iface: A Quantum GIS QGisAppInterface instance.
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: New line before this one

File: safe/gui/tools/print_report_dialog.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+
+        self.unwanted_templates = ['merged_report.qpt', 'infographic.qpt']
+
+        # Load templates from resources...
+        template_dir_path = resources_path('qgis-composer-templates')
+        self.populate_template_combobox(
+            template_dir_path, self.unwanted_templates)
+
+        #  ...and user directory
+        settings = QtCore.QSettings()
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: We are remove the use of QSettings in the code. Can you use safe/utilities/settings.py, thanks. Same for use of QSettings in this PR

File: safe/gui/tools/print_report_dialog.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+                self.create_pdf = False
+                self.open_in_composer()
+            self.dock.hide_busy()
+        except Exception as e:
+            self.dock.hide_busy()
+
+        QtGui.QDialog.accept(self)
+
+    @pyqtSignature('')  # prevents actions being handled twice
+    def on_template_chooser_clicked(self):
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: We also move away from these auto-connect slots. Can you use explicit connections and remove the decorator.

File: safe/gui/tools/print_report_dialog.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+            self.template_combo.setEnabled(True)
+            self.template_path.setEnabled(False)
+            self.template_chooser.setEnabled(False)
+        else:
+            self.template_combo.setEnabled(False)
+            self.template_path.setEnabled(True)
+            self.template_chooser.setEnabled(True)
+
+    @pyqtSlot()
+    @pyqtSignature('bool')  # prevents actions being handled twice
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: Same here

File: safe/gui/widgets/dock.py [hunk 6]
```diff
@@ -745,6 +746,49 @@ def progress_callback(self, current_value, maximum_value, message=None):
...
+            except Exception as e:  # pylint: disable=broad-except
+                error_message = get_error_message(e)
+                send_error_message(self, error_message)
+
+        if self.impact_function:
+            dialog = PrintReportDialog(
+                self.impact_function, self.iface, dock=self, parent=self)
+            dialog.show()
+        else:
+            QtGui.QMessageBox.warning(
```
Comment (Gustry) [inline safe/gui/widgets/dock.py]: safe.utilities.qgis_utilities.display_warning_message_bar thanks less intrusive for UX

https://docs.qgis.org/2.2/en/docs/pyqgis_developer_cookbook/communicating.html#showing-messages-the-qgsmessagebar-class

File: safe/impact_function/impact_function.py [hunk 7]
```diff
@@ -2563,6 +2563,27 @@ def append_ISO19115_keywords(keywords):
...
+
+        :param layer_path: Layer source path.
+        :type layer_path: str
+
+        :return: Vector layer
+        :rtype: QgsVectorLayer
+
+        .. versionadded: 4.3.0
+        """
+        for id, layer in layer_registry.mapLayers().items():
```
Comment (Gustry) [inline safe/impact_function/impact_function.py]: PyCharm is not complaining shadow name with `id` ?

File: safe/impact_function/impact_function.py [hunk 8]
```diff
@@ -2645,37 +2670,34 @@ def load_from_output_metadata(output_metadata):
...
+                ImpactFunction.load_layer_from_registry(
+                    layer_registry, exposure_summary_path))
 
         # aggregate_hazard_impacted
         aggregate_hazard_impacted_path = get_provenance(
             provenance, provenance_layer_aggregate_hazard_impacted)
         if aggregate_hazard_impacted_path:
             impact_function._aggregate_hazard_impacted = load_layer(
                 aggregate_hazard_impacted_path)[0]
-
```
Comment (Gustry) [inline safe/impact_function/impact_function.py]: I added the line to make it more readable last week ;-)

File: safe/impact_function/impact_function.py [hunk 9]
```diff
@@ -2695,6 +2717,8 @@ def load_from_output_metadata(output_metadata):
         else:
             impact_function._crs = None
 
+        impact_function._provenance_ready = True
```
Comment (Gustry) [inline safe/impact_function/impact_function.py]: Where was this statement before?

**Comment by Gustry** [comment]
Hi @myarjunar can you rebase on develop and run `make flake8` on terminal then ?

File: safe/definitions/utilities.py [hunk 10]
```diff
@@ -4,7 +4,9 @@
 from collections import OrderedDict
 from copy import deepcopy
 from os import listdir
-from os.path import join, exists, splitext
+from os.path import join, exists, splitext, split
+from PyQt4 import QtXml
+from PyQt4.QtXml import QDomNode
```
Comment (ismailsunni) [inline safe/definitions/utilities.py]: Perhaps import specific module. You import QDomNode (in QtXml) but also import QtXml above it.

File: safe/definitions/utilities.py [hunk 11]
```diff
@@ -485,6 +487,67 @@ def get_field_groups(layer_purpose, layer_subcategory=None):
...
+    """
+    copy_component = deepcopy(component)
+    template_directory, template_filename = split(template_path)
+    file_name, file_format = splitext(template_filename)
+    if file_format[1:] != (
+            QgisComposerComponentsMetadata.OutputFormat.QPT) or (
+                not exists(template_path)):
+        return copy_component
+
+    from safe.definitions.reports.components import (
```
Comment (ismailsunni) [inline safe/definitions/utilities.py]: Can you put comment why you import inside the method?

File: safe/gui/tools/print_report_dialog.py [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+
+from os.path import join, exists
+from PyQt4 import QtGui, QtCore, QtXml
+from PyQt4.QtCore import pyqtSignature, pyqtSlot
+from qgis.core import QgsApplication, QgsExpressionContextUtils, QgsComposition
+
+from safe import messaging as m
+from safe.common.signals import send_error_message, send_static_message
+from safe.definitions.constants import ANALYSIS_FAILED_BAD_CODE
+from safe.definitions.reports import pdf_product_tag, final_product_tag, \
```
Comment (ismailsunni) [inline safe/gui/tools/print_report_dialog.py]: Avoid backslash

File: safe/gui/tools/print_report_dialog.py [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+    override_component_template, definition, update_template_component)
+from safe.gui.tools.help.impact_report_help import impact_report_help
+from safe.messaging import styles
+from safe.report.impact_report import ImpactReport
+from safe.report.report_metadata import ReportMetadata
+from safe.utilities.i18n import tr
+from safe.utilities.resources import (
+    get_ui_class, resources_path, html_header, html_footer)
+
+__copyright__ = "Copyright 2016, The InaSAFE Project"
```
Comment (ismailsunni) [inline safe/gui/tools/print_report_dialog.py]: 2017

File: safe/gui/tools/print_report_dialog.py (hunk 3 above)
Comment (ismailsunni) [reply safe/gui/tools/print_report_dialog.py]: +1

File: safe/gui/tools/print_report_dialog.py [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+
+        if not path:
+            path = join(QgsApplication.qgisSettingsDirPath(), 'inasafe')
+
+        if exists(path):
+            self.populate_template_combobox(path)
+
+        self.restore_state()
+
+    def populate_template_combobox(self, path, unwanted_templates=[]):
```
Comment (ismailsunni) [inline safe/gui/tools/print_report_dialog.py]: Do not use mutable object (i.e. list) for default parameter. See http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

File: safe/gui/tools/print_report_dialog.py [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+        :type report_path: str
+
+        :param suffix: Expected output product file type (extension).
+        :type suffix: str
+
+        :return: List of absolute path of the output product.
+        :rtype: list
+        """
+        paths = []
+        for c in products:
```
Comment (ismailsunni) [inline safe/gui/tools/print_report_dialog.py]: Avoid single character as variable if possible. It's better to use `for product in products`. And `c` is not obvious, not like `for p in paths`

File: safe/gui/tools/print_report_dialog.py [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+        custom_map_report_metadata = impact_report.metadata
+
+        standard_report_metadata = [
+            standard_impact_report_metadata,
+            standard_map_report_metadata,
+            standard_infographic_report_metadata,
+            custom_map_report_metadata
+        ]
+
+        def retrieve_components(tags):
```
Comment (ismailsunni) [inline safe/gui/tools/print_report_dialog.py]: Docstring here.

File: safe/impact_function/impact_function.py (hunk 9 above)
Comment (myarjunar) [reply safe/impact_function/impact_function.py]: it wasn't anywhere there before. previously  this method doesn't set the provenance for the IF, but I need the IF to have provenances to make t works with the reporting engine, so I added this line as well as the provenance

**Comment by myarjunar** [comment]
thanks for the review @Gustry @ismailsunni !

**Comment by myarjunar** [comment]
should I fix all of this flake8 message? @Gustry 
```
./scripts/create_api_docs.py:131:22: E901 SyntaxError: invalid syntax
./scripts/missing_translations.py:65:22: E901 SyntaxError: invalid syntax
./safe/plugin.py:622:29: E901 SyntaxError: invalid syntax
./safe/messaging/error_message.py:18:1: E402 module level import not at top of file
./safe/messaging/error_message.py:20:1: E402 module level import not at top of file
./safe/messaging/error_message.py:21:1: E402 module level import not at top of file
./safe/messaging/error_message.py:22:1: E402 module level import not at top of file
./safe/messaging/error_message.py:23:1: E402 module level import not at top of file
./safe/messaging/error_message.py:30:1: E402 module level import not at top of file
./safe/messaging/message.py:19:1: E402 module level import not at top of file
./safe/messaging/message.py:20:1: E402 module level import not at top of file
./safe/messaging/message.py:22:1: E402 module level import not at top of file
./safe/messaging/message.py:23:1: E402 module level import not at top of file
./safe/messaging/message.py:24:1: E402 module level import not at top of file
./safe/messaging/item/abstract_list.py:19:1: E402 module level import not at top of file
./safe/messaging/item/abstract_list.py:21:1: E402 module level import not at top of file
./safe/messaging/item/abstract_list.py:23:1: E402 module level import not at top of file
./safe/messaging/item/abstract_list.py:24:1: E402 module level import not at top of file
./safe/messaging/item/abstract_list.py:25:1: E402 module level import not at top of file
./safe/messaging/item/brand.py:20:1: E402 module level import not at top of file
./safe/messaging/item/brand.py:23:1: E402 module level import not at top of file
./safe/messaging/item/bulleted_list.py:18:1: E402 module level import not at top of file
./safe/messaging/item/cell.py:19:1: E402 module level import not at top of file
./safe/messaging/item/cell.py:20:1: E402 module level import not at top of file
./safe/messaging/item/emphasized_text.py:18:1: E402 module level import not at top of file
./safe/messaging/item/heading.py:18:1: E402 module level import not at top of file
./safe/messaging/item/heading.py:20:1: E402 module level import not at top of file
./safe/messaging/item/heading.py:21:1: E402 module level import not at top of file
./safe/messaging/item/horizontal_rule.py:18:1: E402 module level import not at top of file
./safe/messaging/item/image.py:18:1: E402 module level import not at top of file
./safe/messaging/item/important_text.py:18:1: E402 module level import not at top of file
./safe/messaging/item/line_break.py:18:1: E402 module level import not at top of file
./safe/messaging/item/link.py:18:1: E402 module level import not at top of file
./safe/messaging/item/message_element.py:67:37: F821 undefined name 'basestring'
./safe/messaging/item/message_element.py:69:37: F821 undefined name 'long'
./safe/messaging/item/numbered_list.py:18:1: E402 module level import not at top of file
./safe/messaging/item/paragraph.py:18:1: E402 module level import not at top of file
./safe/messaging/item/paragraph.py:19:1: E402 module level import not at top of file
./safe/messaging/item/preformatted_text.py:18:1: E402 module level import not at top of file
./safe/messaging/item/row.py:19:1: E402 module level import not at top of file
./safe/messaging/item/row.py:21:1: E402 module level import not at top of file
./safe/messaging/item/row.py:22:1: E402 module level import not at top of file
./safe/messaging/item/row.py:23:1: E402 module level import not at top of file
./safe/messaging/item/success_paragraph.py:18:1: E402 module level import not at top of file
./safe/messaging/item/text.py:19:1: E402 module level import not at top of file
./safe/messaging/item/text.py:20:1: E402 module level import not at top of file
./safe/messaging/item/text.py:22:1: E402 module level import not at top of file
./safe/messaging/item/text.py:24:1: E402 module level import not at top of file
./safe/messaging/item/text.py:25:1: E402 module level import not at top of file
./safe/messaging/item/text.py:26:1: E402 module level import not at top of file
./safe/report/impact_report.py:73:41: F821 undefined name 'basestring'
./safe/report/impact_report.py:120:29: F821 undefined name 'basestring'
./safe/report/impact_report.py:152:33: F821 undefined name 'basestring'
./safe/report/report_metadata.py:71:36: F821 undefined name 'basestring'
./safe/report/test/hello_world_report.py:25:43: E901 SyntaxError: invalid syntax
./safe/datastore/folder.py:53:30: F821 undefined name 'basestring'
./safe/datastore/geopackage.py:47:30: F821 undefined name 'basestring'
./safe/definitions/test/scripts.py:9:23: E901 SyntaxError: invalid syntax
./safe/common/custom_logging.py:65:26: E901 SyntaxError: invalid syntax
./safe/common/exceptions.py:39:32: F821 undefined name 'unicode'
./safe/common/exceptions.py:55:48: F821 undefined name 'unicode'
./safe/common/exceptions.py:56:28: F821 undefined name 'unicode'
./safe/common/minimum_needs.py:12:1: E402 module level import not at top of file
./safe/common/minimum_needs.py:13:1: E402 module level import not at top of file
./safe/common/minimum_needs.py:14:1: E402 module level import not at top of file
./safe/common/minimum_needs.py:15:1: E402 module level import not at top of file
./safe/common/minimum_needs.py:17:1: E402 module level import not at top of file
./safe/common/signals.py:25:1: E402 module level import not at top of file
./safe/common/signals.py:27:1: E402 module level import not at top of file
./safe/common/signals.py:29:1: E402 module level import not at top of file
./safe/common/utilities.py:128:31: E901 SyntaxError: invalid token
./safe/common/parameters/group_select_parameter_widget.py:157:30: F821 undefined name 'xrange'
./safe/utilities/file_downloader.py:21:1: E402 module level import not at top of file
./safe/utilities/file_downloader.py:25:1: E402 module level import not at top of file
./safe/utilities/file_downloader.py:27:1: E402 module level import not at top of file
./safe/utilities/file_downloader.py:29:1: E402 module level import not at top of file
./safe/utilities/file_downloader.py:31:1: E402 module level import not at top of file
./safe/utilities/file_downloader.py:32:1: E402 module level import not at top of file
./safe/utilities/gis.py:325:15: F821 undefined name 'unicode'
./safe/utilities/keyword_io.py:359:43: F821 undefined name 'basestring'
./safe/utilities/keyword_io.py:465:38: F821 undefined name 'basestring'
./safe/utilities/keyword_io.py:558:43: F821 undefined name 'basestring'
./safe/utilities/pivot_table.py:274:34: F821 undefined name 'xrange'
./safe/utilities/pivot_table.py:275:18: F821 undefined name 'xrange'
./safe/utilities/resources.py:26:10: F821 undefined name 'file'
./safe/utilities/unicode.py:26:36: F821 undefined name 'basestring'
./safe/utilities/unicode.py:44:31: F821 undefined name 'unicode'
./safe/utilities/unicode.py:46:12: F821 undefined name 'unicode'
./safe/utilities/unicode.py:62:31: F821 undefined name 'unicode'
./safe/utilities/unicode.py:80:35: F821 undefined name 'unicode'
./safe/test/debug_helper.py:159:23: E901 SyntaxError: invalid syntax
./safe/test/utilities.py:183:12: F821 undefined name 'file'
./safe/metadata/metadata_db_io.py:153:29: E901 SyntaxError: invalid syntax
./safe/metadata/property/base_property.py:20:1: E402 module level import not at top of file
./safe/metadata/property/base_property.py:22:1: E402 module level import not at top of file
./safe/metadata/property/base_property.py:89:35: F821 undefined name 'unicode'
./safe/metadata/property/base_property.py:146:33: F821 undefined name 'basestring'
./safe/metadata/property/character_string_property.py:29:35: F821 undefined name 'unicode'
./safe/metadata/property/character_string_property.py:33:21: F821 undefined name 'unicode'
./safe/metadata/property/character_string_property.py:51:37: F821 undefined name 'unicode'
./safe/metadata/property/character_string_property.py:53:34: F821 undefined name 'unicode'
./safe/metadata/property/character_string_property.py:54:20: F821 undefined name 'unicode'
./safe/metadata/property/dictionary_property.py:39:34: F821 undefined name 'basestring'
./safe/metadata/property/list_property.py:21:1: E402 module level import not at top of file
./safe/metadata/property/list_property.py:22:1: E402 module level import not at top of file
./safe/metadata/property/list_property.py:24:1: E402 module level import not at top of file
./safe/metadata/property/list_property.py:25:1: E402 module level import not at top of file
./safe/metadata/property/url_property.py:21:1: E402 module level import not at top of file
./safe/metadata/property/url_property.py:23:1: E402 module level import not at top of file
./safe/metadata/property/url_property.py:25:1: E402 module level import not at top of file
./safe/metadata/property/url_property.py:26:1: E402 module level import not at top of file
./safe/metadata/test/scripts.py:26:20: E901 SyntaxError: invalid syntax
./safe/gis/vector/prepare_vector_layer.py:231:32: F821 undefined name 'basestring'
./safe/gis/vector/summary_1_aggregate_hazard.py:182:23: F821 undefined name 'unicode'
./safe/impact_function/postprocessors.py:187:29: F821 undefined name 'reduce'
./safe/impact_function/postprocessors.py:259:44: F821 undefined name 'unicode'
./safe/impact_function/postprocessors.py:329:21: F821 undefined name 'reduce'
./safe/impact_function/postprocessors.py:341:29: F821 undefined name 'reduce'
./safe/gui/gui_utilities.py:49:12: F821 undefined name 'cmp'
./safe/gui/widgets/dock.py:1454:48: F821 undefined name 'unicode'
./safe/gui/widgets/field_mapping_tab.py:234:42: F821 undefined name 'basestring'
./safe/gui/widgets/message_viewer.py:19:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:20:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:25:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:27:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:28:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:29:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:30:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:31:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:32:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:33:1: E402 module level import not at top of file
./safe/gui/widgets/message_viewer.py:319:19: E901 SyntaxError: invalid syntax
./safe/gui/tools/field_mapping_dialog.py:261:29: E901 SyntaxError: invalid syntax
./safe/gui/tools/impact_report_dialog.py:23:1: E402 module level import not at top of file
./safe/gui/tools/impact_report_dialog.py:24:1: E402 module level import not at top of file
./safe/gui/tools/impact_report_dialog.py:26:1: E402 module level import not at top of file
./safe/gui/tools/impact_report_dialog.py:27:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:20:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:21:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:25:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:32:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:36:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:38:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:40:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:43:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:45:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:46:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:49:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:50:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:54:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:57:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:60:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:61:1: E402 module level import not at top of file
./safe/gui/tools/osm_downloader_dialog.py:62:1: E402 module level import not at top of file
./safe/gui/tools/print_report_dialog.py:419:9: F841 local variable 'e' is assigned to but never used
./safe/gui/tools/shake_grid/shake_grid.py:317:26: E901 SyntaxError: invalid syntax
./safe/gui/tools/wizard/step_fc20_hazlayer_from_canvas.py:120:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_fc35_explayer_from_canvas.py:121:25: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_fc55_agglayer_from_canvas.py:123:25: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw00_purpose.py:144:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw05_subcategory.py:142:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw10_hazard_category.py:128:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw20_unit.py:125:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw25_classification.py:146:26: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw30_field.py:143:35: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw30_field.py:275:35: F821 undefined name 'basestring'
./safe/gui/tools/wizard/step_kw30_field.py:279:30: F821 undefined name 'xrange'
./safe/gui/tools/wizard/step_kw33_multi_classifications.py:843:31: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw33_multi_classifications.py:959:53: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw33_multi_classifications.py:993:54: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw40_classify.py:186:24: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw40_classify.py:202:31: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw40_classify.py:292:53: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw40_classify.py:324:54: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw45_inasafe_fields.py:148:45: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw47_default_inasafe_fields.py:143:38: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw49_inasafe_raster_default_values.py:120:38: F821 undefined name 'unicode'
./safe/gui/tools/wizard/step_kw65_summary.py:113:23: F821 undefined name 'file'
./safe/gui/tools/wizard/step_kw65_summary.py:114:23: F821 undefined name 'file'
./safe/gui/tools/wizard/wizard_dialog.py:848:29: E901 SyntaxError: invalid syntax
./safe/gui/tools/batch/batch_dialog.py:21:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:22:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:23:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:24:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:25:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:26:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:28:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:29:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:30:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:37:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:45:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:46:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:47:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:52:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:56:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:60:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:61:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:62:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:63:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:64:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:65:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:66:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:67:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:68:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:70:1: E402 module level import not at top of file
./safe/gui/tools/batch/batch_dialog.py:594:30: E901 SyntaxError: invalid syntax
./safe/gui/tools/help/definitions_help.py:1406:41: F821 undefined name 'unicode'
./safe/gui/tools/help/definitions_help.py:1408:31: F821 undefined name 'unicode'
./safe/gui/tools/help/developer_help.py:243:14: F821 undefined name 'file'
./safe/gui/tools/minimum_needs/needs_manager_dialog.py:609:33: E901 SyntaxError: invalid syntax
./safe/gui/tools/minimum_needs/needs_profile.py:12:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:13:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:15:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:16:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:17:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:19:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:20:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:21:1: E402 module level import not at top of file
./safe/gui/tools/minimum_needs/needs_profile.py:22:1: E402 module level import not at top of file

```

**Comment by Gustry** [comment]
Yes you have to of course :D 

No, seriously, this is in your local. I don't know how you installed flake8 and which packages because travis (and I) as the same results:
```
-----------
Flake8 issues
-----------
3.5.0 (mccabe: 0.6.1, pycodestyle: 2.3.1, pyflakes: 1.6.0) CPython 2.7.6 on Linux
./safe/gui/tools/print_report_dialog.py:216:31: F812 list comprehension redefines 'path' from line 203
./safe/gui/tools/print_report_dialog.py:419:29: F841 local variable 'e' is assigned to but never used
./safe/report/processors/default.py:570:25: E741 ambiguous variable name 'l'
```

**Comment by myarjunar** [comment]
@Gustry :smiley:  , how did you install it btw? I use `apt install flake8`

**Comment by Gustry** [comment]
What are your versions number when you run make flake8 ?

File: safe/gui/tools/print_report_dialog.py [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+        for f in report_files:
+            self.template_combo.addItem(
+                QtCore.QFileInfo(f).baseName(), path + '/' + f)
+
+    def restore_state(self):
+        """Reinstate the options based on the user's stored session info.
+
+        .. versionadded: 4.3.0
+        """
+        settings = QtCore.QSettings()
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: We remove the use of QSettings. use safe/utilities/settings instead

File: safe/gui/tools/print_report_dialog.py [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+        except TypeError:
+            path = ''
+        self.template_path.setText(path)
+
+    def save_state(self):
+        """Store the options into the user's stored session info.
+
+        .. versionadded: 4.3.0
+        """
+        settings = QtCore.QSettings()
```
Comment (Gustry) [inline safe/gui/tools/print_report_dialog.py]: Same here

**Pull request merged**